### PR TITLE
Move getInputs to init_state and update unit tests

### PR DESCRIPTION
**Pull request by rraustad**
Pull request overview
---------------------

Open for discussion.

 - Curious what was needed if we begin moving getInputs to init_state.
 - init_state is called after the input processor so this move should not affect example files as long as the order of calls in init_state allows everything to be read in properly.
 - unit testing does not call init_state so unit tests must set up proper calls to getInputs, GetFluidProperties, etc.
 - I tried a few things in this branch to see what might be needed moving forward.

1) Should init_state be called by unit tests, either by the input processor (process_idf) or the unit test itself. init_state will, IMO, include all calls to getInputs for all objects (and those getInputs will need to return cleanly if no objects exist). Unit tests will not include all objects so this may? not be a big overhead. There is about 250 init_state functions that may eventually call a getInput or other code. Big question is whether to call init_state in process_idf.
2) some unit tests have idf snippets and some do not. I added init_state to process_idf so that writing unit tests may be a little easier. This may add overhead to unit tests but I think that will be minimal. Those unit tests that do not have an idf snippet cannot be set up to automatically call init_state (in SetUp fixture) because if the unit test did have an idf snippet that includes fluid properties those fluid properties need to be called after those inputs are processed. So either the unit test calls init_state in process_idf or the unit test sets up the proper function calls individually.
3) the call added here to GetProjectData reads in objects that were not previously read in before in unit tests so wound up changing some text in error messages or eio reports. There was also a change to some unit tests since now Timestep and CurrentTime had different values. IF I had not added init_state to process_idf then maybe there would not have been as many changes to unit test text messages as seen here but then the unit test itself will be responsible for calling all necessary getInputs (because of 4).
4) I envision that there will no longer be `if (getInputFlag) { getInput(args); getInputFlag = false; }` chunks of code anywhere in EnergyPlus. I removed those calls in FluidProperties to see what was affected. Why burden (might be a strong word) the engine for unit testing? This required that all unit tests that call GetDensityGlycol have a call to GetFluidPropertiesData either in the process_idf or the unit test.
5) I tried both methods in unit tests, calling init_state or setting up proper calls to needed functions, to see what was the best method.

InitializePsychRoutines, OpenOutputFiles and GetProjectData were moved to init_state. GetFluidPropertiesData was already in init_state.

    void init_state(EnergyPlusData &state) override
    {
        Psychrometrics::InitializePsychRoutines(state);
    }

    void init_state(EnergyPlusData &state) override
    {
        SimulationManager::OpenOutputFiles(state);
        SimulationManager::GetProjectData(state);
    }

And added init_state to the API initializer.

    void initializeFunctionalAPI(EnergyPlusState state)
    {
        auto *thisState = reinterpret_cast<EnergyPlus::EnergyPlusData *>(state);
        if (!thisState->dataInputProcessing->inputProcessor) {
            thisState->dataInputProcessing->inputProcessor = EnergyPlus::InputProcessor::factory();
        }
        thisState->init_state(*thisState);
    }

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/SimulationManager.cc [hunk 1]
```diff
@@ -209,12 +209,8 @@ namespace SimulationManager {
              state.dataInputProcessing->inputProcessor->getNumObjectsFound(state, "RunPeriod:CustomRange") > 0 || state.dataSysVars->FullAnnualRun);
         state.dataErrTracking->AskForConnectionsReport = false; // set to false until sizing is finished
 
-        OpenOutputFiles(state);
-
         state.init_state(state);
 
-        GetProjectData(state);
-        Psychrometrics::InitializePsychRoutines(state);
```
Comment (rraustad) [inline src/EnergyPlus/SimulationManager.cc:217]: Fundamental change is merging line 212, 216, and 217 into init_state. The rest is a result of this change.

File: src/EnergyPlus/FluidProperties.cc [hunk 2]
```diff
@@ -162,6 +162,9 @@ namespace FluidProperties {
         //   Array initializer only takes one argument.  std::bind is used to convert the
         //   actual initializer into a function of one argument.
 
+        if (state.dataFluidProps->GetFluidPropertiesData_called) return;
```
Comment (amirroth) [inline src/EnergyPlus/FluidProperties.cc]: Why is this necessary?

File: src/EnergyPlus/Data/EnergyPlusData.cc [hunk 3]
```diff
@@ -570,11 +571,15 @@ void EnergyPlusData::clear_state()
...
+    if (this->init_state_called) return;
+    this->init_state_called = true;
     // The order in which we do this matters.  We're going to try to
     // do this in "topological" order meaning the first to go are the
     // objects that do not reference any other objects, like fluids,
     // schedules, curves, etc.
-    this->dataFluidProps->init_state(state);
+    this->dataSimulationManager->init_state(state); // GetProjectData
+    this->dataFluidProps->init_state(state);        // GetFluidPropertiesData
+    this->dataPsychrometrics->init_state(state);    // InitializePsychRoutines
```
Comment (amirroth) [inline src/EnergyPlus/Data/EnergyPlusData.cc:583]: I was going to suggest that you do this (i.e., start sorting the `init_state` calls) but you are already doing it.

File: tst/EnergyPlus/unit/Autosizing/CoolingWaterDesAirOutletTempSizing.unit.cc [hunk 4]
```diff
@@ -54,11 +54,13 @@
 #include <EnergyPlus/DataHVACGlobals.hh>
 #include <EnergyPlus/DataSizing.hh>
 #include <EnergyPlus/Fans.hh>
+#include <EnergyPlus/FluidProperties.hh>
 
 namespace EnergyPlus {
 
 TEST_F(AutoSizingFixture, CoolingWaterDesAirOutletTempSizingGauntlet)
 {
+    FluidProperties::GetFluidPropertiesData(*state);
```
Comment (amirroth) [inline tst/EnergyPlus/unit/Autosizing/CoolingWaterDesAirOutletTempSizing.unit.cc]: should we do it like this or should we do `state->dataFluidProps->init_state();`?

File: tst/EnergyPlus/unit/BoilerHotWater.unit.cc [hunk 5]
```diff
@@ -71,6 +71,7 @@ using namespace EnergyPlus::Psychrometrics;
 
 TEST_F(EnergyPlusFixture, Boiler_HotWaterSizingTest)
 {
+    state->init_state(*state);
```
Comment (amirroth) [inline tst/EnergyPlus/unit/BoilerHotWater.unit.cc]: Do it like this? Or selectively call `init_state` on the subset of modules you are using in the test?

File: src/EnergyPlus/FluidProperties.cc (hunk 2 above)
Comment (rraustad) [reply src/EnergyPlus/FluidProperties.cc]: For a simulation it's not. But I was worried that unit tests might call GetFluidPropertiesData more than once. If that happens you hit things like this:

    ++state.dataFluidProps->NumOfRefrigerants;

File: tst/EnergyPlus/unit/Autosizing/CoolingWaterDesAirOutletTempSizing.unit.cc (hunk 4 above)
Comment (rraustad) [reply tst/EnergyPlus/unit/Autosizing/CoolingWaterDesAirOutletTempSizing.unit.cc]: Good point, either method will work. I guess it's the developer's preference. On one hand I can see which function is being called, on the other hand developers should know what that means.

File: tst/EnergyPlus/unit/BoilerHotWater.unit.cc (hunk 5 above)
Comment (rraustad) [reply tst/EnergyPlus/unit/BoilerHotWater.unit.cc]: That may be the better approach for unit testing. However, there could be several init_state calls to set everything up.

File: tst/EnergyPlus/unit/Autosizing/CoolingWaterDesAirOutletTempSizing.unit.cc (hunk 4 above)
Comment (rraustad) [reply tst/EnergyPlus/unit/Autosizing/CoolingWaterDesAirOutletTempSizing.unit.cc]: The benefit of your suggestion is not having to include the header if it's missing.

Comment (rraustad) [reply tst/EnergyPlus/unit/Autosizing/CoolingWaterDesAirOutletTempSizing.unit.cc]: Nope, need to include the header either way. So that leaves it to preference.

Comment (amirroth) [reply tst/EnergyPlus/unit/Autosizing/CoolingWaterDesAirOutletTempSizing.unit.cc]: Ok, I saw all of your responses. I'm cool with them. We'll learn as we go.

**Comment by Myoldmopar** [comment]
This is all clean on CI.  I know this is going to bend and flex as more gets added to init state calls, and this is a good step forward.

File: src/EnergyPlus/Data/EnergyPlusData.cc (hunk 3 above)
Comment (Myoldmopar) [reply src/EnergyPlus/Data/EnergyPlusData.cc:583]: I'm not quite 100% convinced we can come up with a single order that works for every single simulation configuration.  I'm happy to be shown that we actually can, but I expected the init calls to be a bit more surgical, with individual inits calling out to child inits.  Anyway, let's see where this goes.

File: src/EnergyPlus/FluidProperties.cc [hunk 6]
```diff
@@ -7572,12 +7489,6 @@ namespace FluidProperties {
         bool LowErrorThisTime = false;
         bool HighErrorThisTime = false;
 
-        // Get the input if we haven't already
-        if (state.dataFluidProps->GetInput) {
-            GetFluidPropertiesData(state);
-            state.dataFluidProps->GetInput = false;
-        }
```
Comment (Myoldmopar) [inline src/EnergyPlus/FluidProperties.cc:7579]: These removals are just beautiful.

File: tst/EnergyPlus/unit/AirTerminalSingleDuct.unit.cc [hunk 7]
```diff
@@ -1008,6 +1008,7 @@ TEST_F(EnergyPlusFixture, SingleDuctVAVReheatAirTerminal_MinFlowTurnDownTest)
     });
 
     ASSERT_TRUE(process_idf(idf_objects));
+    compare_err_stream_substring("", true); // clear idf errors
```
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/AirTerminalSingleDuct.unit.cc:1011]: Hmm, unexpected addition here.  Did new errors get emitted that you need to clear away?  It's not a big deal.  If we need a worker function over on the text fixture to clear away the err buffer, we could add one pretty easily:

```c++
void EnergyPlusFixture::clear_err_stream();  // in EnergyPlusFixture.hh
void EnergyPlusFixture::clear_err_stream()   // in EnergyPlusFixture.cc
{
    this->err_stream->str(std::string());
}
```

Then out here it would just be `clear_err_stream();`

File: tst/EnergyPlus/unit/Autosizing/CoolingWaterDesAirOutletTempSizing.unit.cc [hunk 8]
```diff
@@ -54,11 +54,13 @@
 #include <EnergyPlus/DataHVACGlobals.hh>
 #include <EnergyPlus/DataSizing.hh>
 #include <EnergyPlus/Fans.hh>
+#include <EnergyPlus/FluidProperties.hh>
 
 namespace EnergyPlus {
 
 TEST_F(AutoSizingFixture, CoolingWaterDesAirOutletTempSizingGauntlet)
 {
+    state->dataFluidProps->init_state(*state);
```
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/Autosizing/CoolingWaterDesAirOutletTempSizing.unit.cc:63]: I feel like this should be moved to the AutoSizingFixture base class setUp routine.  If it is something that is going to be called by each child test that uses that fixture, let's just move it out there.  What do you think?

**Comment by rraustad** [comment]
The biggest question I have is whether I should have deleted all those calls to `GetFluidPropertiesData(state);` or left them there as a safety measure for the time being. That would have reduced the effort on unit test updates.

File: tst/EnergyPlus/unit/DOASEffectOnZoneSizing.unit.cc [hunk 9]
```diff
@@ -139,7 +139,7 @@ TEST_F(EnergyPlusFixture, DOASEffectOnZoneSizing_CalcDOASSupCondsForSizing)
 
 TEST_F(EnergyPlusFixture, DOASEffectOnZoneSizing_SizeZoneEquipment)
 {
-
+    state->init_state(*state);
```
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/DOASEffectOnZoneSizing.unit.cc:142]: Hmm...why does this one specific unit test need this added?  Maybe it doesn't since you added it to EnergyPlusFixture?  And we should delete this line?  If not, I'm just curious why it's needed here and not in every other test.

File: tst/EnergyPlus/unit/Furnaces.unit.cc [hunk 10]
```diff
@@ -87,6 +87,7 @@ namespace EnergyPlus {
 TEST_F(EnergyPlusFixture, SetVSHPAirFlowTest_VSFurnaceFlowTest)
 {
 
+    state->init_state(*state);
```
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/Furnaces.unit.cc:90]: Again a little curious/suspicious why just a few specific tests need this added.  I expected either it be added to the fixture base class and no tests would need it, or every test would need it.

File: tst/EnergyPlus/unit/HVACSizingSimulationManager.unit.cc [hunk 11]
```diff
@@ -136,6 +137,8 @@ class HVACSizingSimulationManagerTest : public EnergyPlusFixture
 TEST_F(HVACSizingSimulationManagerTest, WeatherFileDaysTest3)
 {
 
+    FluidProperties::GetFluidPropertiesData(*state);
```
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/HVACSizingSimulationManager.unit.cc:140]: Again, here it would be good to add anything common to all the unit tests into the custom fixture.

**Review by Myoldmopar — COMMENTED** [review]
I have a few comments.  Some of them could be made into action items here, or they could be pushed into a later PR.  Let me know what you think of them and we'll get this moving toward merge either way.

File: tst/EnergyPlus/unit/Autosizing/CoolingWaterDesAirOutletTempSizing.unit.cc (hunk 8 above)
Comment (rraustad) [reply tst/EnergyPlus/unit/Autosizing/CoolingWaterDesAirOutletTempSizing.unit.cc:63]: The problem with fixture vs unit test is whether an idf snippet exists. If idf snippet includes FluidProperties then those properties won't get read in with a fixture call to init_state because that happens before the idf snippet is read in and processed. For this specific autosizing unit test it will likely work as you suggest since there are no idf snippets.

File: tst/EnergyPlus/unit/DOASEffectOnZoneSizing.unit.cc (hunk 9 above)
Comment (rraustad) [reply tst/EnergyPlus/unit/DOASEffectOnZoneSizing.unit.cc:142]: After 2 days of unit test updates and trying different things to find the easiest path I'll have to check if this one and others you point out are sill needed.

File: tst/EnergyPlus/unit/HVACSizingSimulationManager.unit.cc (hunk 11 above)
Comment (rraustad) [reply tst/EnergyPlus/unit/HVACSizingSimulationManager.unit.cc:140]: During unit test updating I couldn't call GetFluidPropertiesData more than once. I fixed that so now that function can be called multiple times without issue. So yes, this one can be moved to the fixture. And regarding my last comment, the init_state can be moved to the fixture as long as the getInputs in init_state can be called multiple times (to update unit tests with idf snippets). I feel like I need a wiki page.

Comment (Myoldmopar) [reply tst/EnergyPlus/unit/HVACSizingSimulationManager.unit.cc:140]: Right.  I see.  It can't be in the fixture.  That's OK.  It's just the same as the process_idf call we have now.  It must be in every unit test that has IDF snippets, and it must be after we declare them.  Maybe we should chat on technicalities to show this off?

Comment (rraustad) [reply tst/EnergyPlus/unit/HVACSizingSimulationManager.unit.cc:140]: A discussion would be good since init_state will affect code and unit testing going forward. I did add init_state to process_idf for unit tests with idf snippets. The question is using init_state in the fixture vs process_idf vs unit test and what that means going forward.

File: tst/EnergyPlus/unit/AirTerminalSingleDuct.unit.cc (hunk 7 above)
Comment (rraustad) [reply tst/EnergyPlus/unit/AirTerminalSingleDuct.unit.cc:1011]: The move of GetProjectData to init_state caused additional text to appear ("Timestep < 4" was 1 of them). I was testing ways to avoid changing the unit test. I think I could have added the *substring method directly in the unit test instead of doing this or what you suggest.

**Comment by Myoldmopar** [comment]
Everything is still happy here with develop pulled in.  I feel like this is a good step in the right direction, and we should merge, with a full understanding that this will get massaged and refactored while we find the right pattern.  Any objections need to come soon, otherwise I'll merge shortly.

**Comment by rraustad** [comment]
I can update unit test init_state logic on the next go around. I think schedules and curve/table objects may be next.

**Comment by Myoldmopar** [comment]
Yep, sounds good.  Just pulled in develop again after the merge from @RKStrand's PR, and this is still good.  Let's merge this.  Thanks @rraustad
